### PR TITLE
IGNITE-17259 Populate ReplicaListener with handlers for RO requests

### PR DESCRIPTION
**Pull request by vldpyatkov**

### CODE REVIEW

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 1]
```diff
@@ -182,12 +189,112 @@ public CompletableFuture<Object> invoke(ReplicaRequest request) {
...
+
+        UUID indexId = indexIdOrDefault(indexScanId/*request.indexToUse()*/);
+
+        if (request.requestType() !=  RequestType.RO_GET) {
+            throw new IgniteInternalException(Replicator.REPLICA_COMMON_ERR,
+                    IgniteStringFormatter.format("Unknown single request [actionType={}]", request.requestType()));
+        }
+
+        //TODO: Use timestamp to find a row id.
+        RowId rowId = rowIdByKey(indexId, searchKey);
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:246]: There may be multiple rowIds. 
k1 -> rowId1
tx begin
  remove k1
  insert k1->rowId2
tx commit

As a result there are two k1 entries in index.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 2]
```diff
@@ -182,12 +189,112 @@ public CompletableFuture<Object> invoke(ReplicaRequest request) {
...
+        return CompletableFuture.completedFuture(result);
+    }
+
+    /**
+     * Processes multiple entries request for read only transaction.
+     *
+     * @param request Read only multiple entries request.
+     * @return Result future.
+     */
+    private CompletableFuture<Object> processReadOnlyMultiEntryAction(ReadOnlyMultiRowReplicaRequest request) {
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:259]: Same as for single key

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 3]
```diff
@@ -182,12 +189,112 @@ public CompletableFuture<Object> invoke(ReplicaRequest request) {
...
+    /**
+     * Resolves a read result to the matched row.
+     * If the result dos not math any row, the method return {@code null}.
+     * TODO: Use a write intent resolution procedure to get a row form {@link ReadResult}.
+     *
+     * @param readResult Read result.
+     * @return {@link BinaryRow} or {@code null}.
+     */
+    private BinaryRow resolveWriteIntent(ReadResult readResult) {
+        if (readResult == null) {
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:291]: Why do we expect null readResult here?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 4]
```diff
@@ -182,12 +189,112 @@ public CompletableFuture<Object> invoke(ReplicaRequest request) {
...
+            RowId rowId = rowIdByKey(indexId, searchKey);
+
+            result.add(rowId != null ? resolveWriteIntent(mvDataStorage.read(rowId, request.timestamp())) : null);
+        }
+
+        return CompletableFuture.completedFuture(result);
+    }
+
+    /**
+     * Resolves a read result to the matched row.
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:283]: Please add more details of how writeIntent resolution works.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 5]
```diff
@@ -182,12 +189,112 @@ public CompletableFuture<Object> invoke(ReplicaRequest request) {
...
+     *
+     * @param readResult Read result.
+     * @return {@link BinaryRow} or {@code null}.
+     */
+    private BinaryRow resolveWriteIntent(ReadResult readResult) {
+        if (readResult == null) {
+            return null;
+        }
+
+        return readResult.binaryRow();
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:295]: I'd rather check if binaryRow is null and throw an IllegalStateException or similar until writeIntent resolution is implemented.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 6]
```diff
@@ -1051,4 +1158,62 @@ private CompletableFuture<Void> ensureReplicaIsPrimary(ReplicaRequest request) {
             return CompletableFuture.completedFuture(null);
         }
     }
+
+    /**
+     * Cursor wrapper to resolve a write intent, which may be required for scan by timestamp.
+     */
+    private class BinaryRowCursorWrapper implements Cursor<BinaryRow> {
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:1165]: Why we need it?

why not to use
```
        for (int i = 0; i < batchCount && cursor.hasNext(); i++) {
            batchRows.add(resolveWriteIntent(cursor.next()));
        }
```
Where cursor will be common PartitionTimestampCursor?

**Review by sanpwc — CHANGES_REQUESTED** [review]

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 7]
```diff
@@ -182,12 +189,112 @@ public CompletableFuture<Object> invoke(ReplicaRequest request) {
...
+                    IgniteStringFormatter.format("Unknown single request [actionType={}]", request.requestType()));
+        }
+
+        ArrayList<BinaryRow> result = new ArrayList<>(keyRows.size());
+
+        for (ByteBuffer searchKey : keyRows) {
+            //TODO: Use timestamp to find a row id.
+            RowId rowId = rowIdByKey(indexId, searchKey);
+
+            result.add(rowId != null ? resolveWriteIntent(mvDataStorage.read(rowId, request.timestamp())) : null);
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:276]: What's the point of adding null in result list?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 8]
```diff
@@ -1665,15 +1665,15 @@ public boolean onUpdate(@NotNull WatchEvent evt) {
                         return true;
                     }
 
-                    int part = extractPartitionNumber(pendingAssignmentsWatchEvent.key());
+                    int partId = extractPartitionNumber(pendingAssignmentsWatchEvent.key());
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:1668]: We are already using the abbreviate in other places to determine partition number.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableMessageGroup.java [hunk 9]
```diff
@@ -58,4 +58,19 @@ public class TableMessageGroup {
      * Message type for {@link org.apache.ignite.internal.table.distributed.message.HasDataResponse}.
      */
     public static final int HAS_DATA_RESPONSE = 6;
+
+    /**
+     * Message type for {@link org.apache.ignite.internal.table.distributed.replication.request.ReadOnlySingleRowReplicaRequest}.
+     */
+    public static final short RO_SINGLE_ROW_REPLICA_REQUEST = 7;
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableMessageGroup.java:65]: It is the same as other constants' declaration here. I do not catch what you mean.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replication/request/ScanRetrieveBatchReplicaRequest.java [hunk 10]
```diff
@@ -33,11 +34,52 @@ public interface ScanRetrieveBatchReplicaRequest extends ReplicaRequest {
...
      *
-     * @return Row filter predicate.
+     * @return Index id.
      */
     @Marshallable
-    Predicate<BinaryRow> rowFilter();
+    UUID indexToUse();
+
+    /**
+     * Gets a key which is used for exact comparison in the index.
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replication/request/ScanRetrieveBatchReplicaRequest.java:45]: Ok, I used these.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 11]
```diff
@@ -182,12 +189,112 @@ public CompletableFuture<Object> invoke(ReplicaRequest request) {
...
+     * @param request Read only retrieve batch request.
+     * @return Result future.
+     */
+    private CompletableFuture<Object> processReadOnlyScanRetrieveBatchAction(ReadOnlyScanRetrieveBatchReplicaRequest request) {
+        UUID txId = request.transactionId();
+        int batchCount = request.batchSize();
+
+        IgniteUuid cursorId = new IgniteUuid(txId, request.scanId());
+
+        PartitionTimestampCursor timestampCursor = mvDataStorage.scan(row -> true, request.timestamp());
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:216]: Done.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java (hunk 6 above)
Comment (vldpyatkov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:1165]: BinaryRowCursorWrapper incapsulates a logic about missing `null` values (which might be resolved for a write intent).

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java (hunk 4 above)
Comment (vldpyatkov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:283]: Added more description.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableMessageGroup.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableMessageGroup.java:65]: I believe that if'll merge main you'll see that all public static part was removed.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:1668]: make sense

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java (hunk 6 above)
Comment (sanpwc) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:1168]: Do you mean one that might be handled as follows:
```
                BinaryRow resolvedReadResult = resolveReadResult(cursor.next());
                if (resolvedReadResult != null) {
                    batchRows.add(resolvedReadResult);
                }
```
?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replication/request/ScanRetrieveBatchReplicaRequest.java [hunk 12]
```diff
@@ -33,11 +35,52 @@ public interface ScanRetrieveBatchReplicaRequest extends ReplicaRequest {
...
+     *
+     * @return Key to search.
+     */
+    @Marshallable
+    BinaryTuple exactKey();
+
+    /**
+     * Gets a lower bound to choose entries from {@link SortedIndexStorage}. Exclusivity is controlled by a {@link
+     * SortedIndexStorage#GREATER_OR_EQUAL} or {@link SortedIndexStorage#GREATER} flag. {@code null} means unbounded.
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replication/request/ScanRetrieveBatchReplicaRequest.java:55]: Mentioning SortedIndexStorage is actually an abstraction leak. I'd expected that similar flags will be introduced in https://issues.apache.org/jira/browse/IGNITE-17748. Please add corresponding todo here, that'll denote that flag links should be changed.

**Review by sanpwc — APPROVED** [review]

**Pull request merged**